### PR TITLE
981 migrate our test cases to nunit with constraints

### LINKED ISSUE #981: Migrate our test cases to NUnit version 3.5
We are currently using Unity package Custom NUnit version 2.0.5, which is based on NUnit version 3.5.

Version 3.5 has a new style for writing tests using constraints. They have some advantages:

- More readable (although also more verbose).
- More informative error messages: a constraint knows what it was comparing, so it reports it.
- Constraints compose; classic methods don't.
- One entry point, and the argument order stops mattering.
- A constraint is a value.
- Delegates, exceptions and polling use the same syntax.

Constraints are where NUnit is going. NUnit 4 moved the classic assertions into NUnit.Framework.Legacy.ClassicAssert. Constraint-model tests carry forward unchanged.

For these reasons, we want our new tests be written with constraints. At the same time, for reasons of consistency, we want our tests in the classic style be migrated to constraints.

About 70% of the assertions are in the classic model. Of the 53 test files that assert anything: 34 pure classic, 5 pure constraint, 14 mixed.

### PR DESCRIPTION
**Pull request by koschke**
Closes #981.

Migrates the whole test suite from NUnit's classic assertion model
(`Assert.AreEqual`, `Assert.IsTrue`, …) to the constraint model
(`Assert.That(actual, Is.…)`), covering 52 files in `Assets/SEETests` and
`Assets/SEEPlayModeTests`: roughly 940 classic assertions plus 187 calls that
passed a bare boolean to `Assert.That` and therefore reported nothing but a line
number when they failed.

Both properties are now enforceable by inspection: neither assembly contains a
classic assertion, a `CollectionAssert`/`StringAssert` call, or an `Assert.That`
without a constraint.

## Approach

The migration is not purely mechanical. Applied throughout:

- The purpose-built constraint is preferred over a literal translation:
  `Is.Empty` instead of `AreEqual(0, x.Count)`, `Is.EquivalentTo` for sets,
  `Does.Exist` for files and directories, `Has.Member`, `Does.StartWith`,
  `Is.InstanceOf<T>`, `Is.Null`.
- `Assert.AreSame` becomes `Is.SameAs`, never `Is.EqualTo`, so reference
  identity stays reference identity.
- `Assert.Throws<T>` becomes `Throws.TypeOf<T>()` rather than
  `Throws.InstanceOf<T>()`, which would also accept subclasses.
- Genuine predicates keep `Is.True`/`Is.False` but gain a hand-written message,
  because the constraint alone reports no more than the classic form did.
- The argument order of every `Assert.AreEqual` was checked individually. Five
  files passed the actual value first, so a mechanical rewrite would have
  swapped expected and actual in the failure report.

## Defects found and fixed along the way

- **`CityCursor` could claim another city's elements.** It decided whether a
  hovered element belongs to this city with `selectedGraph.Equals(city.LoadedGraph)`.
  `Graph.Equals` is value based, so two cities showing two equal but distinct
  graphs both took hover focus. Now compared with `==`, i.e. by identity.
- **`TestReflexionAnalysis` was asserting with Unity's assertions**, not NUnit's:
  the file aliased `Assert` to `UnityEngine.Assertions.Assert`. A violated Unity
  assertion throws `AssertionException`, which NUnit reports as an error rather
  than a failure, and those assertions are conditional on `UNITY_ASSERTIONS` and
  would be compiled away where that symbol is undefined. This affected the three
  derived reflexion fixtures, which call `AssertEventCountEquals`.
- **A missing null check** in `TestJacocoImporter.AddMetricToMethodNode`.
- **A copy-paste error** in `TestPropertyDialog`, which re-checked the game
  object instead of the selector.
- **A message that enumerated a Git query twice** in `TestGitSEE`, on every call
  including passing ones, because the interpolated string was built eagerly.
- Smaller items: a stray UTF-8 BOM in `AddEdgeAction.cs`, two malformed doc
  comments, unused imports, dead code (`SceneQueries.GetGraphs`,
  `SceneQueries.GetRoots`, `GameNodeHierarchy.Check`).

## Verification

The complete **edit-mode** suite was run and reports no errors.

The **play-mode** tests could not be run: the scene they execute in lacks the
required setup. This is a pre-existing condition, unrelated to this change, and
will be handled in its own issue. Consequently the eight touched play-mode files
are compiled but not executed:

`TestDoTween`, `TestKeywordInput`, `TestMenu`, `TestNestedMenu`,
`TestPropertyDialog`, `TestSEEGame`, `TestSimpleMenu`, `TestUI`

Their changes are of the same kind as the verified edit-mode ones, but that is
an argument, not evidence — they should be run once the scene setup is
available.

Three edit-mode fixtures are also excluded from a plain local run, as before
this change: `TestDataPath`, `TestGit` and `TestGitSEE` need network access or
a GitHub access token.

## Known limitation

`TestGraphIO.WriteReadGraphAsync` asserts that the graph written and the graph
read back are equal, but `Graph.Equals` compares only type, `Name` and `Path`,
and the paths are made equal immediately before. The GXL round-trip therefore
verifies nothing about the graph content. The assertions are migrated with their
semantics unchanged and the situation is documented in a comment at the call
site; making the test meaningful requires a structural graph comparison, which
does not exist in the code base yet. Deferred to a separate issue #983 rather than
stretched into this one.

## Note for the reviewer

Two commits on this branch are unrelated to #981 and were produced by tooling:
`4dc265c76f` (URP update to a Dissonance demo material, triggered through Odin
Validator) and `0e7f4a10fd` (settings written by the HighlightPlus import).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by koschke — COMMENTED** [review]
Looks good from my point of view.

File: Assets/SEETests/TestActionStateType.cs [hunk 1]
```diff
@@ -115,10 +118,10 @@ public static IEnumerable<TestCaseData> AllTypeSupplier()
         [Test, TestCaseSource(nameof(AllTypeSupplier))]
         public void TestEquality(AbstractActionStateType type)
         {
-            Assert.IsTrue(type.Equals(type));
-            Assert.IsFalse(type.Equals(null));
-            Assert.AreEqual(1, allRootTypes.AllElements().Where(type.Equals).Count(),
-                            "An ActionStateType must only be equal to itself!");
+            Assert.That(type.Equals(type), Is.True, $"{type.Name} must be equal to itself.");
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestActionStateType.cs:124]: Fixed.

File: Assets/SEETests/TestGraph.cs [hunk 2]
```diff
@@ -510,22 +538,38 @@ private void TestSubGraphBy(Action<GraphElement> makeRelevant, Action<GraphEleme
...
-            Assert.That(BAAAA.IsLeaf);
-            Assert.That(BBA.IsLeaf);
-            Assert.That(BCAA.IsLeaf);
-            Assert.That(BCAB.IsLeaf);
-            Assert.That(BCBA.IsLeaf);
-            Assert.That(BDAA.IsLeaf);
-            Assert.That(C.IsLeaf);
-            Assert.That(E.IsLeaf);
+            Assert.That(BAA.IsRoot, Is.True,
+                        $"{BAA.ID} should be a root.");
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestGraph.cs]: Fixed.

File: Assets/SEETests/TestGraph.cs [hunk 3]
```diff
@@ -510,22 +538,38 @@ private void TestSubGraphBy(Action<GraphElement> makeRelevant, Action<GraphEleme
...
+            Assert.That(BB.IsRoot, Is.True,
+                        $"{BB.ID} should be a root.");
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestGraph.cs]: Fixed.

File: Assets/SEETests/TestGraph.cs [hunk 4]
```diff
@@ -510,22 +538,38 @@ private void TestSubGraphBy(Action<GraphElement> makeRelevant, Action<GraphEleme
...
+            Assert.That(BCA.IsRoot, Is.True,
+                        $"{BCA.ID} should be a root.");
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestGraph.cs]: Fixed.

File: Assets/SEETests/TestGraph.cs [hunk 5]
```diff
@@ -510,22 +538,38 @@ private void TestSubGraphBy(Action<GraphElement> makeRelevant, Action<GraphEleme
...
+            Assert.That(BCBA.IsRoot, Is.True,
+                        $"{BCBA.ID} should be a root.");
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestGraph.cs]: Fixed.

File: Assets/SEETests/TestGraph.cs [hunk 6]
```diff
@@ -510,22 +538,38 @@ private void TestSubGraphBy(Action<GraphElement> makeRelevant, Action<GraphEleme
...
+            Assert.That(BD.IsRoot, Is.True,
+                        $"{BD.ID} should be a root.");
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestGraph.cs]: Fixed.

File: Assets/SEETests/TestGraph.cs [hunk 7]
```diff
@@ -510,22 +538,38 @@ private void TestSubGraphBy(Action<GraphElement> makeRelevant, Action<GraphEleme
...
+            Assert.That(C.IsRoot, Is.True,
+                        $"{C.ID} should be a root.");
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestGraph.cs]: Fixed.

File: Assets/SEETests/TestGraph.cs [hunk 8]
```diff
@@ -510,22 +538,38 @@ private void TestSubGraphBy(Action<GraphElement> makeRelevant, Action<GraphEleme
...
+            Assert.That(DA.IsRoot, Is.True,
+                        $"{DA.ID} should be a root.");
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestGraph.cs]: Fixed.

File: Assets/SEETests/TestGraph.cs [hunk 9]
```diff
@@ -510,22 +538,38 @@ private void TestSubGraphBy(Action<GraphElement> makeRelevant, Action<GraphEleme
...
+            Assert.That(E.IsRoot, Is.True,
+                        $"{E.ID} should be a root.");
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestGraph.cs]: Fixed.

File: Assets/SEETests/TestGraph.cs [hunk 10]
```diff
@@ -510,22 +538,38 @@ private void TestSubGraphBy(Action<GraphElement> makeRelevant, Action<GraphEleme
...
+            Assert.That(BAAAA.IsLeaf, Is.True,
+                        $"{BAAAA.ID} should be a leaf.");
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestGraph.cs]: Fixed.

File: Assets/SEETests/TestGraphProviderIO.cs [hunk 11]
```diff
@@ -17,9 +17,9 @@ internal class TestGraphProviderIO : AbstractTestConfigIO
     {
         public static void AreEqual(MultiGraphPipelineProvider expected, MultiGraphProvider actual)
         {
-            Assert.IsTrue(expected.GetType() == actual.GetType());
+            Assert.That(actual, Is.TypeOf(expected.GetType()));
             MultiGraphPipelineProvider graphPipelineLoaded = actual as MultiGraphPipelineProvider;
-            Assert.AreEqual(expected.Pipeline.Count, graphPipelineLoaded.Pipeline.Count);
+            Assert.That(graphPipelineLoaded.Pipeline.Count, Is.EqualTo(expected.Pipeline.Count));
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestGraphProviderIO.cs:23]: Fixed.

File: Assets/SEETests/TestGraphProviderIO.cs [hunk 12]
```diff
@@ -48,26 +48,26 @@ public static void AreEqual(MultiGraphProvider expected, MultiGraphProvider actu
...
+            Assert.That(gxlLoaded.MaxRevisionsToLoad, Is.EqualTo(expected.MaxRevisionsToLoad));
         }
 
         private static void AreEqualGitEvolutionProviders(GitEvolutionGraphProvider expected, MultiGraphProvider actual)
         {
-            Assert.IsTrue(expected.GetType() == actual.GetType());
+            Assert.That(actual, Is.TypeOf(expected.GetType()));
             GitEvolutionGraphProvider gitLoaded = actual as GitEvolutionGraphProvider;
-            Assert.AreEqual(expected.Date, gitLoaded.Date);
+            Assert.That(gitLoaded.Date, Is.EqualTo(expected.Date));
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestGraphProviderIO.cs:63]: Fixed.

File: Assets/SEETests/TestGraphProviderIO.cs [hunk 13]
```diff
@@ -283,9 +283,9 @@ public void TestNestedPipelineProvider()
         private static void AreEqualSinglePipelineProviders(SingleGraphPipelineProvider expected,
                                                             SingleGraphProvider actual)
         {
-            Assert.IsTrue(expected.GetType() == actual.GetType());
+            Assert.That(actual, Is.TypeOf(expected.GetType()));
             SingleGraphPipelineProvider graphPipelineLoaded = actual as SingleGraphPipelineProvider;
-            Assert.AreEqual(expected.Pipeline.Count, graphPipelineLoaded.Pipeline.Count);
+            Assert.That(graphPipelineLoaded.Pipeline.Count, Is.EqualTo(expected.Pipeline.Count));
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestGraphProviderIO.cs:291]: Fixed.

File: Assets/SEETests/TestGraphProviderIO.cs [hunk 14]
```diff
@@ -315,9 +315,9 @@ public void TestAllBranchGitSingleProvider()
         private void AreEqualAllBranchGitSingleProvider(GitBranchesGraphProvider expected,
                                                         SingleGraphProvider actual)
         {
-            Assert.AreEqual(expected.GetType(), actual.GetType());
+            Assert.That(actual, Is.TypeOf(expected.GetType()));
             GitBranchesGraphProvider gitBranchesLoaded = actual as GitBranchesGraphProvider;
-            Assert.AreEqual(gitBranchesLoaded.SimplifyGraph, expected.SimplifyGraph);
+            Assert.That(gitBranchesLoaded.SimplifyGraph, Is.EqualTo(expected.SimplifyGraph));
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestGraphProviderIO.cs:324]: Fixed.

File: Assets/SEETests/TestActionStateType.cs [hunk 15]
```diff
@@ -115,10 +118,11 @@ public static IEnumerable<TestCaseData> AllTypeSupplier()
         [Test, TestCaseSource(nameof(AllTypeSupplier))]
         public void TestEquality(AbstractActionStateType type)
         {
-            Assert.IsTrue(type.Equals(type));
-            Assert.IsFalse(type.Equals(null));
-            Assert.AreEqual(1, allRootTypes.AllElements().Where(type.Equals).Count(),
-                            "An ActionStateType must only be equal to itself!");
+            Assert.That(type, Is.Not.Null, "Type must not be null.");
+            Assert.That(type.Equals(type), Is.True, $"{type.Name} must be equal to itself.");
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestActionStateType.cs:124]: We redefined `Equals`. The purpose of this test is to check whether the redefinition yields true when comparing the type with itself.

File: Assets/SEETests/TestActionStateType.cs [hunk 16]
```diff
@@ -115,10 +118,11 @@ public static IEnumerable<TestCaseData> AllTypeSupplier()
...
+            Assert.That(type.Equals(null), Is.False, $"{type.Name} must not be equal to null.");
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestActionStateType.cs:127]: We redefined `Equals`. The purpose of this test is to check whether the redefinition yields true when comparing against null.

File: Assets/SEETests/TestLocalGitRepository.cs [hunk 17]
```diff
@@ -74,13 +87,15 @@ public static void SetUp()
...
-            Assert.IsTrue(File.Exists(Path.Combine(originalRepoPath, "firstFile.cs")));
-            Assert.IsTrue(File.Exists(Path.Combine(cloneRepoPath, "firstFile.cs")));
+            Assert.That(new DirectoryInfo(originalRepoPath), Does.Exist);
+            Assert.That(new DirectoryInfo(cloneRepoPath), Does.Exist);
+            Assert.That(Repository.IsValid(originalRepoPath), Is.True,
+                        $"{originalRepoPath} is not a valid Git repository.");
+            Assert.That(Repository.IsValid(cloneRepoPath), Is.True,
+                        $"{cloneRepoPath} is not a valid Git repository.");
+
+            Assert.That(new FileInfo(Path.Combine(originalRepoPath, firstFile)), Does.Exist);
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestLocalGitRepository.cs:104]: `firstFile` is `private const string firstFile = "firstFile.cs"`, hence, relative. The problem flagged cannot occur.

File: Assets/SEETests/TestLocalGitRepository.cs [hunk 18]
```diff
@@ -74,13 +87,15 @@ public static void SetUp()
...
+            Assert.That(new FileInfo(Path.Combine(cloneRepoPath, firstFile)), Does.Exist);
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestLocalGitRepository.cs:105]: `firstFile` is `private const string firstFile = "firstFile.cs"`, hence, relative. The problem flagged cannot occur.

**Comment by koschke** [comment]
All newer issues raised by Copilot addressed. Some of them were false positives. But I fixed the problem in `WriteFile` (not actually flagged by Copilot) where it could (in the future) occur.

File: Assets/SEE/Game/CityRendering/GameNodeHierarchy.cs [hunk 19]
```diff
@@ -56,37 +56,6 @@ static void CollectNodes(GameObject root, IDictionary<Node, GameObject> nodeMap)
...
-                if (graphs.Count > 1)
-                {
-                    Debug.LogError("There are nodes from different graphs in the same game-node hierarchy!\n");
-                    foreach (GameObject go in nodeMap.Values)
-                    {
-                        Node node = go.GetNode();
-                        Debug.LogWarning($"Node {node.ID} contained in graph {node.ItsGraph.Name} from file {node.ItsGraph.Path}\n");
-                    }
-                }
-                Assert.AreEqual(1, graphs.Count);
```
Comment (Cyclone1337) [inline Assets/SEE/Game/CityRendering/GameNodeHierarchy.cs:88]: Since this removes the last use of UnityEngine.Assertions.Assert from this file, using UnityEngine.Assertions; at the top can be removed as well.

File: Assets/SEEPlayModeTests/TestUI.cs [hunk 20]
```diff
@@ -17,7 +17,7 @@ internal abstract class TestUI
         /// <summary>
         /// Setup for a test.
         /// The playmode will be entered.
-        /// The <see cref="SceneSettings.InputType"/> will be <see cref="SEE.GO.PlayerInputType.DesktopPlayer"/>.
+        /// The <see cref="User.UserSettings.Instance.InputType"/> will be <see cref="SEE.GO.PlayerInputType.DesktopPlayer"/>.
```
Comment (Cyclone1337) [inline Assets/SEEPlayModeTests/TestUI.cs]: `User.UserSettings.Instance.InputType` is a runtime access path rather than a valid member reference for `cref`. Could we reference `User.UserSettings.InputType` and keep the current wording as the link text, e.g. `<see cref="User.UserSettings.InputType">UserSettings.Instance.InputType</see>`? That keeps it navigable while still documenting the actual access path. Or simply use `<see cref="User.UserSettings.InputType"/>`.

File: Assets/SEETests/UI/TestMenuEntry.cs [hunk 21]
```diff
@@ -49,18 +48,14 @@ public void TestConstructorDefault()
...
             void Action() => testItems.Add(1);
             MenuEntry entry = CreateMenuEntry(Action, "Test");
-            Assert.AreEqual(null, entry.Description);
-            Assert.AreEqual("Test", entry.Title);
-            Assert.AreEqual(true, entry.Enabled);
-            Assert.AreEqual(' ', entry.Icon);
-            Assert.AreEqual(default(Color), entry.EntryColor);
-#if INCLUDE_STEAM_VR
-
-            Assert.AreNotEqual(default(Color), entry.DisabledColor, "Entry color must differ from disabled color!");
```
Comment (Cyclone1337) [inline Assets/SEETests/UI/TestMenuEntry.cs:59]: Why wasn't the `DisabledColor` assertion migrated as well? Also, is there a specific reason why this check was guarded by `INCLUDE_STEAM_VR` in the first place? `DisabledColor` seems to be a regular `MenuEntry` property and is also used outside the SteamVR-specific code.

File: Assets/SEETests/TestLocalGitRepository.cs [hunk 22]
```diff
@@ -33,11 +33,18 @@ internal class TestLocalGitRepository : TestGitRepository
         ///
         /// Then a git commit is made
         /// </summary>
-        /// <param name="path">The path of the file</param>
+        /// <param name="repo">The repository to commit the file to</param>
```
Comment (Cyclone1337) [inline Assets/SEETests/TestLocalGitRepository.cs]: According to our coding guidelines, the parameter descriptions are missing a period at the end.

File: Assets/SEETests/TestRanges.cs [hunk 23]
```diff
@@ -82,99 +86,99 @@ private static char ComparisonToSymbol(int comparison) =>
...
-            Assert.IsFalse(LargeLineRange.Contains(301, 0));
-            Assert.IsFalse(LargeLineRange.Contains(301, 50));
+            Assert.That(OneFullLine.Contains(0, 0), Is.False, $"{OneFullLine} must not contain 0:0.");
+            Assert.That(OneFullLine.Contains(0, 5), Is.False, $"{OneFullLine} must not contain 0:5.");
+            Assert.That(OneFullLine.Contains(1, 0), Is.True, $"{OneFullLine} must contain 1:0.");
+            Assert.That(OneFullLine.Contains(1, 1), Is.True, $"{OneFullLine} must contain 1:1.");
+            Assert.That(OneFullLine.Contains(1, 10), Is.True, $"{OneFullLine} must contain 1:10.");
+            Assert.That(OneFullLine.Contains(2, 0), Is.False, $"{OneFullLine} must not contain 2:0.");
+
+            Assert.That(OneFullLine.Contains(0, 0), Is.False, $"{OneFullLine} must not contain 0:0.");
```
Comment (Cyclone1337) [inline Assets/SEETests/TestRanges.cs]: This block appears to test `TwoFullLines`, but the first two and the last assertion still use `OneFullLine`. Should these three assertions use `TwoFullLines` as well? Otherwise the boundary cases for `TwoFullLines` remain untested.

File: Assets/SEETests/TestGraphCloning.cs [hunk 24]
```diff
@@ -127,23 +127,28 @@ public void TestCloneGraph()
...
-
-            // all cloned nodes must be in the cloned graph
+            Assert.That(clone.Path, Is.EqualTo(original.Path));
+            Assert.That(clone.Name, Is.EqualTo(original.Name));
+            Assert.That(clone.NodeCount, Is.EqualTo(original.NodeCount));
+            Assert.That(clone.EdgeCount, Is.EqualTo(original.EdgeCount));
+
+            // All cloned nodes must be in the cloned graph.
+            // Note: The graph is compared by identity (Is.SameAs), not by equality
+            // (Is.EqualTo). Graph.Equals considers only Name and Path, both of which a
```
Comment (Cyclone1337) [inline Assets/SEETests/TestGraphCloning.cs]: Tiny documentation nit: `Graph.Equals` is a little pickier than the comment gives it credit for, it also checks the concrete graph type in addition to `Name` and `Path`. 😄

File: Assets/SEETests/TestGraphIO.cs [hunk 25]
```diff
@@ -180,20 +180,23 @@ private static async UniTask WriteReadGraphAsync(string basename, Graph outGraph
...
                 Graph backupGraph = await LoadGraphAsync(backupPath);
                 // The path of backupGraph will be backupFilename.
-                Assert.AreEqual(backupPath.Path, backupGraph.Path);
+                Assert.That(backupGraph.Path, Is.EqualTo(backupPath.Path));
                 // For the comparison, we need to reset the path.
                 backupGraph.Path = inGraph.Path = outGraph.Path;
 
-                Assert.AreEqual(outGraph, inGraph);
-                Assert.AreEqual(backupGraph, inGraph);
+                // Note that Graph.Equals compares only the type, Name and Path of a graph.
```
Comment (Cyclone1337) [inline Assets/SEETests/TestGraphIO.cs:196]: Since the comment points out that these assertions only compare type, name, and path, would it be worth checking the actual graph contents here as well?

**Review by Cyclone1337 — APPROVED** [review]
Thank you for the pull request! Overall, the changes look good to me. I only found a few minor issues and questions during the review, mostly regarding test coverage, documentation, and some small cleanups. Nothing major from my side.

File: Assets/SEETests/TestActionStateType.cs [hunk 26]
```diff
@@ -115,10 +118,15 @@ public static IEnumerable<TestCaseData> AllTypeSupplier()
...
-            Assert.IsTrue(type.Equals(type));
-            Assert.IsFalse(type.Equals(null));
-            Assert.AreEqual(1, allRootTypes.AllElements().Where(type.Equals).Count(),
-                            "An ActionStateType must only be equal to itself!");
+            Assert.That(type, Is.Not.Null, "Type must not be null.");
+            // We redefined `Equals`. The purpose of this test is to check
+            // whether the redefinition yields true when comparing the type with itself.
+            Assert.That(type.Equals(type), Is.True, $"{type.Name} must be equal to itself.");
+            // We redefined `Equals`. The purpose of this test is to check whether
+            // the redefinition yields true when comparing against null.
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestActionStateType.cs]: Good catch.

File: Assets/SEE/Game/CityRendering/GameNodeHierarchy.cs (hunk 19 above)
Comment (koschke) [reply Assets/SEE/Game/CityRendering/GameNodeHierarchy.cs:88]: Fixed.

File: Assets/SEEPlayModeTests/TestUI.cs (hunk 20 above)
Comment (koschke) [reply Assets/SEEPlayModeTests/TestUI.cs]: I followed your first advice. Fixed now.

File: Assets/SEETests/UI/TestMenuEntry.cs (hunk 21 above)
Comment (koschke) [reply Assets/SEETests/UI/TestMenuEntry.cs:59]: The preprocessor statement on `INCLUDE_STEAM_VR` was once necessary when we used SteamVR. This component is obsolete for a long time and superseded by Unity's own VR framework. It can be safely deleted.

Comment (koschke) [reply Assets/SEETests/UI/TestMenuEntry.cs:59]: Actually, I do not know why this assert was guarded by the preprocessor condition. It wasn't written by me.

I re-added it as:

`Assert.That(entry.DisabledColor, Is.Not.EqualTo(default(Color)), "Entry color must differ from disabled color!");`

File: Assets/SEETests/TestRanges.cs (hunk 23 above)
Comment (koschke) [reply Assets/SEETests/TestRanges.cs]: You caught a copy&paste mistake! I fixed it.

File: Assets/SEETests/TestLocalGitRepository.cs (hunk 22 above)
Comment (koschke) [reply Assets/SEETests/TestLocalGitRepository.cs]: Fixed.

File: Assets/SEETests/TestGraphCloning.cs (hunk 24 above)
Comment (koschke) [reply Assets/SEETests/TestGraphCloning.cs]: True. I fixed the comment.

File: Assets/SEETests/TestGraphIO.cs (hunk 25 above)
Comment (koschke) [reply Assets/SEETests/TestGraphIO.cs:196]: True, I added a content comparison assisted by Claude Code. It's fixed now.

File: Assets/SEETests/TestActionStateType.cs [hunk 27]
```diff
@@ -115,10 +118,15 @@ public static IEnumerable<TestCaseData> AllTypeSupplier()
...
         public void TestEquality(AbstractActionStateType type)
         {
-            Assert.IsTrue(type.Equals(type));
-            Assert.IsFalse(type.Equals(null));
-            Assert.AreEqual(1, allRootTypes.AllElements().Where(type.Equals).Count(),
-                            "An ActionStateType must only be equal to itself!");
+            Assert.That(type, Is.Not.Null, "Type must not be null.");
+            // We redefined `Equals`. The purpose of this test is to check
+            // whether the redefinition yields true when comparing the type with itself.
+            Assert.That(type.Equals(type), Is.True, $"{type.Name} must be equal to itself.");
```
[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestActionStateType.cs:124]: Duplicate. Was already dismissed.

[the remark this answers is not in the document]
Comment (koschke) [reply Assets/SEETests/TestActionStateType.cs:124]: Duplicate. Was already dismissed.

**Comment by koschke** [comment]
All issues are addressed now. Can be merged.
